### PR TITLE
Feature/128 full crud operation

### PR DESCRIPTION
**Pull request by drekoumis**
Change API use CRUD operation for delete and rename. Create API for rename wf and job . Create some API test for delete and rename wf. Change return API codes.

### CODE REVIEW

File: panoptes/routes/api.py [hunk 1]
```diff
@@ -1,36 +1,35 @@
...
 
 @routes.route('/api/workflow/<workflow_id>', methods=['GET'])
 def get_workflow_by_id(workflow_id):
     workflows = get_db_workflows_by_id(workflow_id)
     if workflows:
-        return jsonify({'workflow': workflows.get_workflow()})
+        return jsonify({'workflow': workflows.get_workflow()}), 200
     else:
-        return jsonify({'msg': "Workflow not found"})
+        return jsonify({'msg': "Workflow not found"}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:32]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 2]
```diff
@@ -39,11 +38,11 @@ def get_jobs_of_workflow(workflow_id):
     if workflows:
         jobs = [j.get_job_json() for j in get_db_jobs(workflows.id)]
         return jsonify({'jobs': jobs,
-                        'count': len(jobs)})
+                        'count': len(jobs)}), 200
     else:
         return jsonify({'msg': 'Workflow not found',
                         'jobs': [],
-                        'count': 0})
+                        'count': 0}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:45]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 3]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
     if workflows:
         job = get_db_job_by_id(workflows.id, job_id)
         if job:
-            return jsonify({'jobs': job.get_job_json()})
+            return jsonify({'jobs': job.get_job_json()}), 200
         else:
-            return jsonify({'msg': 'Job not found'})
+            return jsonify({'msg': 'Job not found'}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:64]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 4]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
     else:
         return jsonify({'msg': 'Workflow not found',
                         'jobs': [],
-                        'count': 0})
+                        'count': 0}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:68]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 5]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
+    if 'name' not in data or len(data['name'])<1:
+        return jsonify({'msg': "Bad Request"}), 400
+    workflows = get_db_workflows_by_id(workflow_id)
+    if workflows:
+        if rename_db_wf(workflow_id, data['name']):
+            return jsonify({'workflow': workflows.get_workflow()}), 200
+        else:
+            return jsonify({'error': 500, 'msg': 'Database error'}), 500
+    else:
+        return jsonify({'msg': "Workflow not found"}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:83]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 6]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
+
+
+@routes.route('/api/workflow/<workflow_id>/job/<job_id>', methods=['PUT'])
+def rename_job_of_workflow(workflow_id, job_id):
+    data = request.json
+    if 'name' not in data or len(data['name'])<1:
+        return jsonify({'msg': "Bad Request"}), 400
```
Comment (dafentoulis) [inline panoptes/routes/api.py:90]: same here for 400, it is already a BAD REQUEST

File: panoptes/routes/api.py [hunk 7]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
+        job = get_db_job_by_id(workflows.id, job_id)
+        if job:
+            old_job_name = job.name
+            rename = rename_db_job(workflow_id, job_id, data['name'])
+            if rename:
+                return jsonify({'jobs': job.get_job_json()}), 200
+            else:
+                return jsonify({'error': 500, 'msg': 'Database error'}), 500
+        else:
+            return jsonify({'msg': 'Job not found'}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:102]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 8]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
+    else:
+        return jsonify({'msg': 'Workflow not found',
+                        'jobs': [],
+                        'count': 0}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:106]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 9]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
 def set_db_delete(workflow_id):
     if(get_db_workflows_by_id(workflow_id) is None):
-        return jsonify({'error': 404, 'msg': 'Unable to delete Workflow ' + workflow_id + 
-                        '. Please check if workflow ' + workflow_id + ' exists.'})
-    elif(get_db_workflows_by_status(workflow_id)=='Running'):
-        return jsonify({'error': 'Delete Rejected', 'msg': 'You cannot delete Running Workflow '})
-    else:        
-        delete=delete_db_wf(workflow_id)
+        return jsonify({'msg': 'Unable to delete Workflow ' + workflow_id +
+                        '. Please check if workflow ' + workflow_id + ' exists.'}), 404
```
Comment (dafentoulis) [inline panoptes/routes/api.py:113]: I think message not needed here, 404 is self-explanatory

File: panoptes/routes/api.py [hunk 10]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
+    elif(get_db_workflows_by_status(workflow_id) == 'Running'):
+        return jsonify({'msg': 'You cannot delete Running Workflow '}), 403
```
Comment (dafentoulis) [inline panoptes/routes/api.py:115]: Change message to "Workflow is running"

File: panoptes/routes/api.py [hunk 11]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
+    else:
+        delete = delete_db_wf(workflow_id)
         if delete:
-          return jsonify({'msg': "Delete Complete Correctly ",'Workflow': workflow_id})
+            return jsonify({'msg': "Delete Complete Correctly ", 'Workflow': workflow_id}), 200
         else:
-             return jsonify({'error': 404, 'msg': 'Database error'})
+            return jsonify({'msg': 'Database error'}), 500
```
Comment (dafentoulis) [inline panoptes/routes/api.py:121]: Message is needed here, but it should describe the problem better

File: panoptes/routes/api.py [hunk 12]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
 def set_whole_db_delete():
     if get_db_table_is_empty('Workflows'):
-        return jsonify({'error': 404, 'msg': 'Database is empty'})
+        return jsonify({'error': 510, 'msg': 'Database is empty'}), 510
     else:
-        delete=delete_whole_db()
+        delete = delete_whole_db()
         if delete:
-            return jsonify({'msg': 'Database clean up is complete'})
+            return jsonify({'msg': 'Database clean up is complete'}), 200
```
Comment (dafentoulis) [inline panoptes/routes/api.py:131]: No message needed, just status code 204.

File: panoptes/routes/api.py [hunk 13]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
         else:
-            return jsonify({'error': 404, 'msg': 'Database error'})
\ No newline at end of file
+            return jsonify({'error': 500, 'msg': 'Database error'}), 500
```
Comment (dafentoulis) [inline panoptes/routes/api.py:133]: Message is needed here, but it should describe the problem better

File: panoptes/routes/api.py [hunk 14]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
-             return jsonify({'error': 404, 'msg': 'Database error'})
+            return jsonify({'msg': 'Database error'}), 500
 
 
-@routes.route('/api/clean-up-database', methods=['GET'])
+@routes.route('/api/workflows/all', methods=['DELETE'])
 def set_whole_db_delete():
     if get_db_table_is_empty('Workflows'):
-        return jsonify({'error': 404, 'msg': 'Database is empty'})
+        return jsonify({'error': 510, 'msg': 'Database is empty'}), 510
```
Comment (dafentoulis) [inline panoptes/routes/api.py:127]: No message needed, status should be 410

File: panoptes/routes/api.py [hunk 15]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
-        delete=delete_db_wf(workflow_id)
+        return jsonify({'msg': 'Unable to delete Workflow ' + workflow_id +
+                        '. Please check if workflow ' + workflow_id + ' exists.'}), 404
+    elif(get_db_workflows_by_status(workflow_id) == 'Running'):
+        return jsonify({'msg': 'You cannot delete Running Workflow '}), 403
+    else:
+        delete = delete_db_wf(workflow_id)
         if delete:
-          return jsonify({'msg': "Delete Complete Correctly ",'Workflow': workflow_id})
+            return jsonify({'msg': "Delete Complete Correctly ", 'Workflow': workflow_id}), 200
```
Comment (dafentoulis) [inline panoptes/routes/api.py:119]: No message needed, just status code 204.

File: panoptes/routes/api.py [hunk 16]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
-                        'count': 0})
+                        'count': 0}), 404
 
 
-@routes.route('/api/delete/<workflow_id>', methods=['GET'])
+@routes.route('/api/workflow/<workflow_id>', methods=['PUT'])
+def rename_workflow_by_id(workflow_id):
+    data = request.json
+    if 'name' not in data or len(data['name'])<1:
+        return jsonify({'msg': "Bad Request"}), 400
```
Comment (dafentoulis) [inline panoptes/routes/api.py:75]: same here for 400, it is already a BAD REQUEST

File: panoptes/routes/api.py [hunk 17]
```diff
@@ -60,37 +59,75 @@ def get_job_of_workflow(workflow_id, job_id):
...
+    workflows = get_db_workflows_by_id(workflow_id)
+    if workflows:
+        if rename_db_wf(workflow_id, data['name']):
+            return jsonify({'workflow': workflows.get_workflow()}), 200
+        else:
+            return jsonify({'error': 500, 'msg': 'Database error'}), 500
```
Comment (dafentoulis) [inline panoptes/routes/api.py:81]: Message is needed here, but it should describe the problem better

**Review by dafentoulis — APPROVED** [review]
Really good job @drekoumis!
I added API related changes, I think @agardelakos  can have a better look at the python - testing part.

**Review by dafentoulis — APPROVED** [review]

**Review by fgypas — CHANGES_REQUESTED** [review]
It seems that the API changed. @drekoumis Can you please, document the changes in the API section of the README.md?

**Review by fgypas — APPROVED** [review]
LGTM !!!

**Pull request merged**